File: evaluation/components/TPS25047_R1_R2_R3.m
% Octave script to solve for R2 and R3 in voltage divider protection circuit
% Equations:
% VIN_UV = (VUVLO_R * (R1+R2+R3)/(R2+R3))
% VIN_OV = (VOV_R * (R1+R2+R3)/(R3))

fprintf('TPS25047 Resistor Divider Calculator\n');
fprintf('========================================\n');

% Fixed reference voltages
VUVLO_R = 1.2;
VOV_R = 1.2;

%% ========================================================================
%% CASE 1: 12V System
%% ========================================================================
fprintf('\n\nCASE 1: 12V System\n');

VIN = 12;
VIN_UV = 10.8;
VIN_OV = 13.2;
R1 = 470e3;

fprintf('\n========================================\n');
fprintf('Input Parameters:\n');
fprintf('VIN = %.2f V\n', VIN);
fprintf('VIN_UV = %.2f V\n', VIN_UV);
fprintf('VIN_OV = %.2f V\n', VIN_OV);
fprintf('R1 = %.2f kOhm\n', R1/1e3);
fprintf('VUVLO_R = %.2f V\n', VUVLO_R);
fprintf('VOV_R = %.2f V\n', VOV_R);

% Solve using matrix algebra
% From equation 1: VIN_UV = (VUVLO_R * (R1 + R2 + R3) / (R2 + R3))
% Rearranged: (VIN_UV - VUVLO_R)*R2 + (VIN_UV - VUVLO_R)*R3 = VUVLO_R*R1
%
% From equation 2: VIN_OV = (VOV_R * (R1 + R2 + R3) / (R3))
% Rearranged: -VOV_R*R2 + (VIN_OV - VOV_R)*R3 = VOV_R*R1
%
% Matrix form: A*x = b where x = [R2; R3]
A = [(VIN_UV - VUVLO_R), (VIN_UV - VUVLO_R);
     -VOV_R, (VIN_OV - VOV_R)];

b = [VUVLO_R * R1;
     VOV_R * R1];

% Solve for [R2; R3]
x = A \ b;

R2 = x(1);
R3 = x(2);

fprintf('\n----------------------------------------\n');
fprintf('Solution:\n');
fprintf('R2 = %.2f Ohm = %.2f kOhm\n', R2, R2/1e3);
fprintf('R3 = %.2f Ohm = %.2f kOhm\n', R3, R3/1e3);

% Check if resistances are positive (physically realizable)
if R2 <= 0 || R3 <= 0
    fprintf('\nWARNING: Negative resistance detected! Not physically realizable.\n');
end

% Verify the solution
fprintf('\n----------------------------------------\n');
fprintf('Verification:\n');
VIN_UV_check = (VUVLO_R * (R1+R2+R3)/(R2+R3));
VIN_OV_check = (VOV_R * (R1+R2+R3)/(R3));
fprintf('VIN_UV from equation: %.4f V (should be %.4f V)\n', VIN_UV_check, VIN_UV);
fprintf('VIN_OV from equation: %.4f V (should be %.4f V)\n', VIN_OV_check, VIN_OV);

% Calculate errors
error_UV = abs(VIN_UV_check - VIN_UV);
error_OV = abs(VIN_OV_check - VIN_OV);
fprintf('Error UV: %.6f V\n', error_UV);
fprintf('Error OV: %.6f V\n', error_OV);

if error_UV < 1e-6 && error_OV < 1e-6
    fprintf('\nSolution verified successfully!\n');
else
    fprintf('\nWARNING: Solution verification failed!\n');
end

%% ========================================================================
%% CASE 2: 5V System
%% ========================================================================
fprintf('\n\nCASE 2: 5V System\n');

VIN = 5;
VIN_UV = 4.6;
VIN_OV = 5.4;
R1 = 470e3;

fprintf('\n========================================\n');
fprintf('Input Parameters:\n');
fprintf('VIN = %.2f V\n', VIN);
fprintf('VIN_UV = %.2f V\n', VIN_UV);
fprintf('VIN_OV = %.2f V\n', VIN_OV);
fprintf('R1 = %.2f kOhm\n', R1/1e3);
fprintf('VUVLO_R = %.2f V\n', VUVLO_R);
fprintf('VOV_R = %.2f V\n', VOV_R);

% Solve using matrix algebra
% From equation 1: VIN_UV = (VUVLO_R * (R1 + R2 + R3) / (R2 + R3))
% Rearranged: (VIN_UV - VUVLO_R)*R2 + (VIN_UV - VUVLO_R)*R3 = VUVLO_R*R1
%
% From equation 2: VIN_OV = (VOV_R * (R1 + R2 + R3) / (R3))
% Rearranged: -VOV_R*R2 + (VIN_OV - VOV_R)*R3 = VOV_R*R1
%
% Matrix form: A*x = b where x = [R2; R3]
A = [(VIN_UV - VUVLO_R), (VIN_UV - VUVLO_R);
     -VOV_R, (VIN_OV - VOV_R)];

b = [VUVLO_R * R1;
     VOV_R * R1];

% Solve for [R2; R3]
x = A \ b;

R2 = x(1);
R3 = x(2);

fprintf('\n----------------------------------------\n');
fprintf('Solution:\n');
fprintf('R2 = %.2f Ohm = %.2f kOhm\n', R2, R2/1e3);
fprintf('R3 = %.2f Ohm = %.2f kOhm\n', R3, R3/1e3);

% Check if resistances are positive (physically realizable)
if R2 <= 0 || R3 <= 0
    fprintf('\nWARNING: Negative resistance detected! Not physically realizable.\n');
end

% Verify the solution
fprintf('\n----------------------------------------\n');
fprintf('Verification:\n');
VIN_UV_check = (VUVLO_R * (R1+R2+R3)/(R2+R3));
VIN_OV_check = (VOV_R * (R1+R2+R3)/(R3));
fprintf('VIN_UV from equation: %.4f V (should be %.4f V)\n', VIN_UV_check, VIN_UV);
fprintf('VIN_OV from equation: %.4f V (should be %.4f V)\n', VIN_OV_check, VIN_OV);

% Calculate errors
error_UV = abs(VIN_UV_check - VIN_UV);
error_OV = abs(VIN_OV_check - VIN_OV);
fprintf('Error UV: %.6f V\n', error_UV);
fprintf('Error OV: %.6f V\n', error_OV);

if error_UV < 1e-6 && error_OV < 1e-6
    fprintf('\nSolution verified successfully!\n');
else
    fprintf('\nWARNING: Solution verification failed!\n');
end

File: evaluation/components/TPS25047/TPS25047_R1_R2_R3.py
#!/usr/bin/env python3
"""Resistor divider calculator for TPS25047.

Converted from the Octave script TPS25047_R1_R2_R3.m.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Case:
    name: str
    vin: float
    vin_uv: float
    vin_ov: float
    r1_ohm: float


@dataclass(frozen=True)
class Solution:
    r2_ohm: float
    r3_ohm: float


def solve_r2_r3(case: Case, vuvlo_r: float, vov_r: float) -> Solution:
    """Solve the 2x2 system for R2 and R3 using closed-form algebra."""
    a = case.vin_uv - vuvlo_r
    b = case.vin_uv - vuvlo_r
    c = -vov_r
    d = case.vin_ov - vov_r

    b1 = vuvlo_r * case.r1_ohm
    b2 = vov_r * case.r1_ohm

    det = a * d - b * c
    if det == 0:
        raise ValueError("Singular matrix: cannot solve for R2 and R3.")

    r2 = (b1 * d - b * b2) / det
    r3 = (a * b2 - b1 * c) / det

    return Solution(r2_ohm=r2, r3_ohm=r3)


def verify(
    case: Case, sol: Solution, vuvlo_r: float, vov_r: float
) -> tuple[float, float]:
    vin_uv_check = (
        vuvlo_r * (case.r1_ohm + sol.r2_ohm + sol.r3_ohm) / (sol.r2_ohm + sol.r3_ohm)
    )
    vin_ov_check = vov_r * (case.r1_ohm + sol.r2_ohm + sol.r3_ohm) / sol.r3_ohm
    return vin_uv_check, vin_ov_check


def run_case(case: Case) -> None:
    vuvlo_r = 1.2
    vov_r = 1.2

    print("\n\n{}".format(case.name))
    print("\n========================================")
    print("Input Parameters:")
    print("VIN = {:.2f} V".format(case.vin))
    print("VIN_UV = {:.2f} V".format(case.vin_uv))
    print("VIN_OV = {:.2f} V".format(case.vin_ov))
    print("R1 = {:.2f} kOhm".format(case.r1_ohm / 1e3))
    print("VUVLO_R = {:.2f} V".format(vuvlo_r))
    print("VOV_R = {:.2f} V".format(vov_r))

    sol = solve_r2_r3(case, vuvlo_r, vov_r)

    print("\n----------------------------------------")
    print("Solution:")
    print("R2 = {:.2f} Ohm = {:.2f} kOhm".format(sol.r2_ohm, sol.r2_ohm / 1e3))
    print("R3 = {:.2f} Ohm = {:.2f} kOhm".format(sol.r3_ohm, sol.r3_ohm / 1e3))

    if sol.r2_ohm <= 0 or sol.r3_ohm <= 0:
        print("\nWARNING: Negative resistance detected! Not physically realizable.")

    print("\n----------------------------------------")
    print("Verification:")
    vin_uv_check, vin_ov_check = verify(case, sol, vuvlo_r, vov_r)
    print(
        "VIN_UV from equation: {:.4f} V (should be {:.4f} V)".format(
            vin_uv_check, case.vin_uv
        )
    )
    print(
        "VIN_OV from equation: {:.4f} V (should be {:.4f} V)".format(
            vin_ov_check, case.vin_ov
        )
    )

    error_uv = abs(vin_uv_check - case.vin_uv)
    error_ov = abs(vin_ov_check - case.vin_ov)
    print("Error UV: {:.6f} V".format(error_uv))
    print("Error OV: {:.6f} V".format(error_ov))

    if error_uv < 1e-6 and error_ov < 1e-6:
        print("\nSolution verified successfully!")
    else:
        print("\nWARNING: Solution verification failed!")


def main() -> None:
    print("TPS25047 Resistor Divider Calculator")
    print("========================================")

    run_case(
        Case(
            name="12V Reference Design",
            vin=12.0,
            vin_uv=10.8,
            vin_ov=13.2,
            r1_ohm=470e3,
        )
    )
    run_case(
        Case(
            name="5V Octohub4",
            vin=5.0,
            vin_uv=4.6,
            vin_ov=5.4,
            r1_ohm=470e3,
        )
    )


if __name__ == "__main__":
    main()
